File: app/src/main/java/apincer/android/mmate/ui/SettingsActivity.java
package apincer.android.mmate.ui;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.Window;
import android.widget.ImageView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.core.view.WindowInsetsControllerCompat;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.PreferenceManager;

import apincer.android.mmate.R;
import apincer.android.mmate.service.MediaServerManager;

public class SettingsActivity extends AppCompatActivity {
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES);
        super.onCreate(savedInstanceState);

        // Enable Edge-to-Edge
        WindowCompat.setDecorFitsSystemWindows(getWindow(), false);

        setContentView(R.layout.activity_fragement);

        ImageView blurBackground = findViewById(R.id.fragment_background_blur);
        // We'll use a standard placeholder or representative art for settings
        // To keep it simple, we don't inject tagRepos here, but just find art if needed.
        
        // set status bar color to black
        Window window = getWindow();
        WindowInsetsControllerCompat insetsController = WindowCompat.getInsetsController(window, window.getDecorView());
        insetsController.setAppearanceLightStatusBars(false);

        androidx.appcompat.widget.Toolbar toolbar = findViewById(R.id.settings_toolbar);
        ViewCompat.setOnApplyWindowInsetsListener(toolbar, (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(v.getPaddingLeft(), systemBars.top, v.getPaddingRight(), v.getPaddingBottom());
            return insets;
        });
        setSupportActionBar(toolbar);

        android.widget.FrameLayout mainContent = findViewById(R.id.main_content);
        ViewCompat.setOnApplyWindowInsetsListener(mainContent, (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(v.getPaddingLeft(), v.getPaddingTop(), v.getPaddingRight(), systemBars.bottom);
            return insets;
        });
        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle("Settings Master");
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setHomeButtonEnabled(true);
        }
        getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.main_content, new SettingsFragment())
                .commit();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {//do whatever
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    public static class SettingsFragment extends PreferenceFragmentCompat implements SharedPreferences.OnSharedPreferenceChangeListener {

        @Override
        public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
            setPreferencesFromResource(R.xml.settings, rootKey);
        }

        @Override
        public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, @Nullable String key) {
            if (apincer.music.core.Constants.PREF_SERVER_ENGINE.equals(key)) {
                if (getActivity() != null) {
                    // Trigger server restart via MediaServerManager when engine setting changes
                    MediaServerManager manager = new MediaServerManager(getActivity().getApplicationContext());
                    manager.doBindService();
                    manager.restartServer();
                }
            }
        }

        @Override
        public void onResume() {
            super.onResume();
            PreferenceManager.getDefaultSharedPreferences(getContext()).registerOnSharedPreferenceChangeListener(this);
        }

        @Override
        public void onPause() {
            super.onPause();
            PreferenceManager.getDefaultSharedPreferences(getContext()).unregisterOnSharedPreferenceChangeListener(this);
        }
    }
}

File: app/src/main/java/apincer/android/mmate/ui/SettingsActivity.kt
package apincer.android.mmate.ui

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.appcompat.app.AppCompatDelegate
import androidx.compose.runtime.*
import androidx.core.view.WindowCompat
import androidx.preference.PreferenceManager
import apincer.android.mmate.service.MediaServerManager
import apincer.android.mmate.ui.compose.MusicMateTheme
import apincer.android.mmate.ui.compose.SettingsScreen
import apincer.music.core.Constants
import apincer.music.core.Settings

class SettingsActivity : ComponentActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)

        val prefs = PreferenceManager.getDefaultSharedPreferences(this)

        setContent {
            var serverEngine by remember {
                mutableStateOf(prefs.getString(Constants.PREF_SERVER_ENGINE, "httpcore") ?: "httpcore")
            }
            var showStorageSpace by remember {
                mutableStateOf(Settings.isShowStorageSpace(this@SettingsActivity))
            }
            var prefixTrackNumber by remember {
                mutableStateOf(Settings.isShowTrackNumber(this@SettingsActivity))
            }
            var listFollowsNowPlaying by remember {
                mutableStateOf(Settings.isListFollowNowPlaying(this@SettingsActivity))
            }
            var artistAwareSimilarSongs by remember {
                mutableStateOf(Settings.isArtistAwareSimilarSongs(this@SettingsActivity))
            }

            MusicMateTheme {
                SettingsScreen(
                    serverEngine = serverEngine,
                    onServerEngineChange = { newEngine ->
                        serverEngine = newEngine
                        prefs.edit().putString(Constants.PREF_SERVER_ENGINE, newEngine).apply()
                        val manager = MediaServerManager(applicationContext)
                        manager.doBindService()
                        manager.restartServer()
                    },
                    showStorageSpace = showStorageSpace,
                    onShowStorageSpaceChange = { checked ->
                        showStorageSpace = checked
                        prefs.edit().putBoolean(Constants.PREF_SHOW_STORAGE_SPACE, checked).apply()
                    },
                    prefixTrackNumber = prefixTrackNumber,
                    onPrefixTrackNumberChange = { checked ->
                        prefixTrackNumber = checked
                        prefs.edit().putBoolean(Constants.PREF_PREFIX_TRACK_NUMBER_ON_TITLE, checked).apply()
                    },
                    listFollowsNowPlaying = listFollowsNowPlaying,
                    onListFollowsNowPlayingChange = { checked ->
                        listFollowsNowPlaying = checked
                        prefs.edit().putBoolean(Constants.PREF_LIST_FOLLOW_NOW_PLAYING, checked).apply()
                    },
                    artistAwareSimilarSongs = artistAwareSimilarSongs,
                    onArtistAwareSimilarSongsChange = { checked ->
                        artistAwareSimilarSongs = checked
                        prefs.edit().putBoolean(Constants.PREF_ARTIST_AWARE_SIMILAR_SONGS, checked).apply()
                    },
                    onBackClick = { finish() }
                )
            }
        }
    }
}
